Populate destination name when destination server is specified

adriananeci opened the pull request:
<!--
Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.
-->

Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [ ] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

<!-- Please see [Contribution FAQs](https://argo-cd.readthedocs.io/en/latest/developer-guide/faq/) if you have questions about your pull-request. -->

mtougeron commented on util/argo/argo.go:
```diff
@@ -508,6 +508,17 @@ func ValidateDestination(ctx context.Context, dest *argoappv1.ApplicationDestina
+			if serverName == "" {
+				return fmt.Errorf("application references destination cluster %s which does not exist", dest.Server)
+			}
+			dest.SetInferredServer(serverName)
```
This still needs to be set to dest.Server doesn't it? Or else the dest.Server field will become the name of the server which will break things. right?

adriananeci replied:
Yup, noticed that earlier, should be addressed now in https://github.com/argoproj/argo-cd/pull/21063/commits/8eb2e4ad8a42bd155d9af543a2157bc618c42c56

mtougeron commented on util/argo/argo.go:
```diff
@@ -508,6 +508,17 @@ func ValidateDestination(ctx context.Context, dest *argoappv1.ApplicationDestina
+			if serverName == "" {
+				return fmt.Errorf("application references destination cluster %s which does not exist", dest.Server)
+			}
+			dest.SetInferredServer(serverName)
```
Somewhere you need to also set the dest.Name to the serverName value or else it won't be set.

adriananeci replied:
Yup, addressed in via https://github.com/argoproj/argo-cd/pull/21063/commits/8eb2e4ad8a42bd155d9af543a2157bc618c42c56

adriananeci commented on applicationset/utils/clusterUtils.go:
```diff
@@ -40,6 +40,8 @@ var (
 	}
 	initLocalCluster sync.Once
+
+	ListClustersFunc = ListClusters
```
Yup, these are needed only for testing purposes. I can try using the k8s fake client instead and see how things go.

adriananeci replied:
Adjusted the appset tests and remove the `ListClustersFunc` global var

adriananeci commented on server/application/application.go:
```diff
@@ -369,7 +369,13 @@ func (s *Server) Create(ctx context.Context, q *application.ApplicationCreateReq
 		return nil, status.Errorf(codes.Internal, "unable to check existing application details (%s): %v", appNs, err)
 	}
-	equalSpecs := reflect.DeepEqual(existing.Spec, a.Spec) &&
+
+	if err := argo.ValidateDestination(ctx, &existing.Spec.Destination, s.db); err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "application destination spec for %s is invalid: %s", existing.Name, err.Error())
+	}
+
+	equalSpecs := existing.Spec.Destination.Equals(a.Spec.Destination) &&
+		reflect.DeepEqual(existing.Spec, a.Spec) &&
```
> if the user does change from name to server in the spec that's still a valid change to the spec and we should be updating the resource in the cluster?

True, but as any other change done to the app spec it needs an upsert since the spec was changed, isn't it?

> Plus as far as I can see in your code if there's already an app with just destination name set then ValidateDestination() will also add the server component to the actual spec of the App so if the user sends the exact same app with the name set the DeepEqual check will fail because your ValidateDestination just update the spec with the missing field effectively making it not equal spec even though it was equal spec

This is not true since the `ValidateDestination()` is also called for the application from the request via `validateAndNormalizeApp` function call, see line 337 above.

adriananeci commented:
Raised https://github.com/argoproj/argo-cd/pull/21176

adriananeci commented:
Hmm, if there is a race condition I think it should have been there for SetInferredServer method too. Maybe there weren't enough tests to validate it until now?

crenshaw-dev commented:
The failure is on an old test, `TestMaxPodLogsRender`:

```
=== Failed
=== FAIL: server/application TestMaxPodLogsRender/PodLogs#02 (0.25s)
time="2024-12-14T05:21:17Z" level=info msg="Starting configmap/secret informers"
time="2024-12-14T05:21:17Z" level=info msg="Configmap/secret informer synced"
==================
WARNING: DATA RACE
Write at 0x00c007c8a138 by goroutine 3293:
  github.com/argoproj/argo-cd/v2/pkg/apis/application/v1alpha1.(*ApplicationDestination).SetInferredName()
      /home/runner/work/argo-cd/argo-cd/pkg/apis/application/v1alpha1/types.go:3335 +0x584
  github.com/argoproj/argo-cd/v2/util/argo.ValidateDestination()
      /home/runner/work/argo-cd/argo-cd/util/argo/argo.go:522 +0x55f
  github.com/argoproj/argo-cd/v2/server/application.(*Server).getApplicationClusterConfig()
      /home/runner/work/argo-cd/argo-cd/server/application/application.go:1303 +0xfc
  github.com/argoproj/argo-cd/v2/server/application.(*Server).PodLogs()
      /home/runner/work/argo-cd/argo-cd/server/application/application.go:1740 +0x1170
  github.com/argoproj/argo-cd/v2/server/application.TestMaxPodLogsRender.func3()
      /home/runner/work/argo-cd/argo-cd/server/application/application_test.go:2217 +0x1bb
  testing.tRunner()
      /opt/hostedtoolcache/go/1.23.3/x64/src/testing/testing.go:1690 +0x226
  testing.(*T).Run.gowrap1()
      /opt/hostedtoolcache/go/1.23.3/x64/src/testing/testing.go:1743 +0x44

Previous read at 0x00c007c8a138 by goroutine 3291:
  github.com/argoproj/argo-cd/v2/pkg/apis/application/v1alpha1.(*ApplicationDestination).String()
      /home/runner/work/argo-cd/argo-cd/pkg/apis/application/v1alpha1/generated.pb.go:18720 +0x184
  github.com/argoproj/argo-cd/v2/pkg/apis/application/v1alpha1.(*ApplicationSpec).String()
      /home/runner/work/argo-cd/argo-cd/pkg/apis/application/v1alpha1/generated.pb.go:19268 +0x572
  github.com/argoproj/argo-cd/v2/pkg/apis/application/v1alpha1.(*Application).String()
      /home/runner/work/argo-cd/argo-cd/pkg/apis/application/v1alpha1/generated.pb.go:18694 +0x11a
  github.com/argoproj/argo-cd/v2/pkg/apis/application/v1alpha1.(*Application).String()
      /home/runner/work/argo-cd/argo-cd/pkg/apis/application/v1alpha1/generated.pb.go:18693 +0xb9
  fmt.(*pp).handleMethods()
      /opt/hostedtoolcache/go/1.23.3/x64/src/fmt/print.go:673 +0x6ea
  fmt.(*pp).printArg()
      /opt/hostedtoolcache/go/1.23.3/x64/src/fmt/print.go:756 +0xb4b
  fmt.(*pp).doPrintf()
      /opt/hostedtoolcache/go/1.23.3/x64/src/fmt/print.go:1173 +0x10ce
  fmt.Sprintf()
      /opt/hostedtoolcache/go/1.23.3/x64/src/fmt/print.go:239 +0x5c
  github.com/stretchr/testify/mock.Arguments.Diff()
      /home/runner/go/pkg/mod/github.com/stretchr/testify@v1.10.0/mock/mock.go:965 +0x1ac
  github.com/stretchr/testify/mock.(*Mock).findExpectedCall()
      /home/runner/go/pkg/mod/github.com/stretchr/testify@v1.10.0/mock/mock.go:383 +0x147
  github.com/stretchr/testify/mock.(*Mock).MethodCalled()
      /home/runner/go/pkg/mod/github.com/stretchr/testify@v1.10.0/mock/mock.go:491 +0xac
  github.com/stretchr/testify/mock.(*Mock).Called()
      /home/runner/go/pkg/mod/github.com/stretchr/testify@v1.10.0/mock/mock.go:481 +0x195
  github.com/argoproj/argo-cd/v2/server/application/mocks.(*Broadcaster).OnAdd()
      /home/runner/work/argo-cd/argo-cd/server/application/mocks/Broadcaster.go:17 +0x136
  k8s.io/client-go/tools/cache.(*processorListener).run.func1()
      /home/runner/go/pkg/mod/k8s.io/client-go@v0.31.0/tools/cache/shared_informer.go:978 +0x287
  k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1()
      /home/runner/go/pkg/mod/k8s.io/apimachinery@v0.31.2/pkg/util/wait/backoff.go:226 +0x41
  k8s.io/apimachinery/pkg/util/wait.BackoffUntil()
      /home/runner/go/pkg/mod/k8s.io/apimachinery@v0.31.2/pkg/util/wait/backoff.go:227 +0xc4
  k8s.io/apimachinery/pkg/util/wait.JitterUntil()
      /home/runner/go/pkg/mod/k8s.io/apimachinery@v0.31.2/pkg/util/wait/backoff.go:204 +0x10a
  k8s.io/apimachinery/pkg/util/wait.Until()
      /home/runner/go/pkg/mod/k8s.io/apimachinery@v0.31.2/pkg/util/wait/backoff.go:161 +0x9b
  k8s.io/client-go/tools/cache.(*processorListener).run()
      /home/runner/go/pkg/mod/k8s.io/client-go@v0.31.0/tools/cache/shared_informer.go:972 +0x38
  k8s.io/client-go/tools/cache.(*processorListener).run-fm()
      <autogenerated>:1 +0x33
  k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1()
      /home/runner/go/pkg/mod/k8s.io/apimachinery@v0.31.2/pkg/util/wait/wait.go:72 +0x86

Goroutine 3293 (running) created at:
  testing.(*T).Run()
      /opt/hostedtoolcache/go/1.23.3/x64/src/testing/testing.go:1743 +0x825
  github.com/argoproj/argo-cd/v2/server/application.TestMaxPodLogsRender()
      /home/runner/work/argo-cd/argo-cd/server/application/application_test.go:2216 +0x4d5
  testing.tRunner()
      /opt/hostedtoolcache/go/1.23.3/x64/src/testing/testing.go:1690 +0x226
  testing.(*T).Run.gowrap1()
      /opt/hostedtoolcache/go/1.23.3/x64/src/testing/testing.go:1743 +0x44

Goroutine 3291 (running) created at:
  k8s.io/apimachinery/pkg/util/wait.(*Group).Start()
      /home/runner/go/pkg/mod/k8s.io/apimachinery@v0.31.2/pkg/util/wait/wait.go:70 +0xe4
  k8s.io/client-go/tools/cache.(*sharedProcessor).addListener()
      /home/runner/go/pkg/mod/k8s.io/client-go@v0.31.0/tools/cache/shared_informer.go:748 +0x224
  k8s.io/client-go/tools/cache.(*sharedIndexInformer).AddEventHandlerWithResyncPeriod()
      /home/runner/go/pkg/mod/k8s.io/client-go@v0.31.0/tools/cache/shared_informer.go:623 +0x770
  k8s.io/client-go/tools/cache.(*sharedIndexInformer).AddEventHandler()
      /home/runner/go/pkg/mod/k8s.io/client-go@v0.31.0/tools/cache/shared_informer.go:561 +0x51
  github.com/argoproj/argo-cd/v2/server/application.NewServer()
      /home/runner/work/argo-cd/argo-cd/server/application/application.go:120 +0x1a5
  github.com/argoproj/argo-cd/v2/server/application.newTestAppServerWithEnforcerConfigure()
      /home/runner/work/argo-cd/argo-cd/server/application/application_test.go:297 +0x2b71
  github.com/argoproj/argo-cd/v2/server/application.createAppServerWithMaxLodLogs()
      /home/runner/work/argo-cd/argo-cd/server/application/application_test.go:2280 +0x9fd
  github.com/argoproj/argo-cd/v2/server/application.TestMaxPodLogsRender()
      /home/runner/work/argo-cd/argo-cd/server/application/application_test.go:2214 +0x38d
  testing.tRunner()
      /opt/hostedtoolcache/go/1.23.3/x64/src/testing/testing.go:1690 +0x226
  testing.(*T).Run.gowrap1()
      /opt/hostedtoolcache/go/1.23.3/x64/src/testing/testing.go:1743 +0x44
==================
    testing.go:1399: race detected during execution of test

=== FAIL: server/application TestMaxPodLogsRender (3.41s)
```

adriananeci commented:
> The failure is on an old test, `TestMaxPodLogsRender`

That's true, but the race condition was always there. It just didn't have a test associated with to validate this scenario.
To confirm it was there even before this PR, I just did a simple test like:
- reverted this PR (`git reset --hard 8dbddb101ac8fa509f9ce857afcc7a57deccd59a`)
- adjusted the `testApp` (https://github.com/argoproj/argo-cd/blob/master/server/application/application_test.go#L2264 from `testApp := newTestApp` to `testApp := newTestAppWithDestName`
- ran the test using `go test -race -run TestMaxPodLogsRender ./server/application -count=1` and got:
```
WARNING: DATA RACE
Write at 0x00c000b09518 by goroutine 400:
  github.com/argoproj/argo-cd/v2/pkg/apis/application/v1alpha1.(*ApplicationDestination).SetInferredServer()
      /Users/aneci/gits/argo-cd/pkg/apis/application/v1alpha1/types.go:3316 +0x134
  github.com/argoproj/argo-cd/v2/util/argo.ValidateDestination()
      /Users/aneci/gits/argo-cd/util/argo/argo.go:507 +0x120
  github.com/argoproj/argo-cd/v2/server/application.(*Server).getApplicationClusterConfig()
      /Users/aneci/gits/argo-cd/server/application/application.go:1297 +0x70
  github.com/argoproj/argo-cd/v2/server/application.(*Server).PodLogs()
      /Users/aneci/gits/argo-cd/server/application/application.go:1734 +0xb10
  github.com/argoproj/argo-cd/v2/server/application.TestMaxPodLogsRender.func2()
      /Users/aneci/gits/argo-cd/server/application/application_test.go:2204 +0x190
  testing.tRunner()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1690 +0x184
  testing.(*T).Run.gowrap1()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1743 +0x40

Previous read at 0x00c000b09518 by goroutine 398:
  github.com/argoproj/argo-cd/v2/pkg/apis/application/v1alpha1.(*ApplicationDestination).String()
      /Users/aneci/gits/argo-cd/pkg/apis/application/v1alpha1/generated.pb.go:18718 +0x38
  github.com/argoproj/argo-cd/v2/pkg/apis/application/v1alpha1.(*ApplicationSpec).String()
      /Users/aneci/gits/argo-cd/pkg/apis/application/v1alpha1/generated.pb.go:19268 +0x450
  github.com/argoproj/argo-cd/v2/pkg/apis/application/v1alpha1.(*Application).String()
      /Users/aneci/gits/argo-cd/pkg/apis/application/v1alpha1/generated.pb.go:18694 +0xd4
  github.com/argoproj/argo-cd/v2/pkg/apis/application/v1alpha1.(*Application).String()
      /Users/aneci/gits/argo-cd/pkg/apis/application/v1alpha1/generated.pb.go:18693 +0x80
  fmt.(*pp).handleMethods()
      /opt/homebrew/opt/go/libexec/src/fmt/print.go:673 +0x54c
  fmt.(*pp).printArg()
      /opt/homebrew/opt/go/libexec/src/fmt/print.go:756 +0x874
  fmt.(*pp).doPrintf()
      /opt/homebrew/opt/go/libexec/src/fmt/print.go:1173 +0xc58
  fmt.Sprintf()
      /opt/homebrew/opt/go/libexec/src/fmt/print.go:239 +0x50
  github.com/stretchr/testify/mock.Arguments.Diff()
      /Users/aneci/gits/argo-cd/vendor/github.com/stretchr/testify/mock/mock.go:965 +0x140
  github.com/stretchr/testify/mock.(*Mock).findExpectedCall()
      /Users/aneci/gits/argo-cd/vendor/github.com/stretchr/testify/mock/mock.go:383 +0x118
  github.com/stretchr/testify/mock.(*Mock).MethodCalled()
      /Users/aneci/gits/argo-cd/vendor/github.com/stretchr/testify/mock/mock.go:491 +0x6c
  github.com/stretchr/testify/mock.(*Mock).Called()
      /Users/aneci/gits/argo-cd/vendor/github.com/stretchr/testify/mock/mock.go:481 +0x150
  github.com/argoproj/argo-cd/v2/server/application/mocks.(*Broadcaster).OnAdd()
      /Users/aneci/gits/argo-cd/server/application/mocks/Broadcaster.go:17 +0xe0
  k8s.io/client-go/tools/cache.(*processorListener).run.func1()
      /Users/aneci/gits/argo-cd/vendor/k8s.io/client-go/tools/cache/shared_informer.go:978 +0x244
  k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1()
      /Users/aneci/gits/argo-cd/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:226 +0x48
  k8s.io/apimachinery/pkg/util/wait.BackoffUntil()
      /Users/aneci/gits/argo-cd/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:227 +0x94
  k8s.io/apimachinery/pkg/util/wait.JitterUntil()
      /Users/aneci/gits/argo-cd/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:204 +0xe4
  k8s.io/apimachinery/pkg/util/wait.Until()
      /Users/aneci/gits/argo-cd/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:161 +0x88
  k8s.io/client-go/tools/cache.(*processorListener).run()
      /Users/aneci/gits/argo-cd/vendor/k8s.io/client-go/tools/cache/shared_informer.go:972 +0x38
  k8s.io/client-go/tools/cache.(*processorListener).run-fm()
      <autogenerated>:1 +0x34
  k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1()
      /Users/aneci/gits/argo-cd/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:72 +0x80

Goroutine 400 (running) created at:
  testing.(*T).Run()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1743 +0x5e0
  github.com/argoproj/argo-cd/v2/server/application.TestMaxPodLogsRender()
      /Users/aneci/gits/argo-cd/server/application/application_test.go:2203 +0x324
  testing.tRunner()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1690 +0x184
  testing.(*T).Run.gowrap1()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1743 +0x40

Goroutine 398 (running) created at:
  k8s.io/apimachinery/pkg/util/wait.(*Group).Start()
      /Users/aneci/gits/argo-cd/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:70 +0xd4
  k8s.io/client-go/tools/cache.(*sharedProcessor).addListener()
      /Users/aneci/gits/argo-cd/vendor/k8s.io/client-go/tools/cache/shared_informer.go:748 +0x1a0
  k8s.io/client-go/tools/cache.(*sharedIndexInformer).AddEventHandlerWithResyncPeriod()
      /Users/aneci/gits/argo-cd/vendor/k8s.io/client-go/tools/cache/shared_informer.go:623 +0x550
  k8s.io/client-go/tools/cache.(*sharedIndexInformer).AddEventHandler()
      /Users/aneci/gits/argo-cd/vendor/k8s.io/client-go/tools/cache/shared_informer.go:561 +0x4c
  github.com/argoproj/argo-cd/v2/server/application.NewServer()
      /Users/aneci/gits/argo-cd/server/application/application.go:120 +0xf8
  github.com/argoproj/argo-cd/v2/server/application.newTestAppServerWithEnforcerConfigure()
      /Users/aneci/gits/argo-cd/server/application/application_test.go:297 +0x2160
  github.com/argoproj/argo-cd/v2/server/application.newTestAppServer()
      /Users/aneci/gits/argo-cd/server/application/application_test.go:139 +0x5c
  github.com/argoproj/argo-cd/v2/server/application.createAppServerWithMaxLodLogs()
      /Users/aneci/gits/argo-cd/server/application/application_test.go:2283 +0x754
  github.com/argoproj/argo-cd/v2/server/application.TestMaxPodLogsRender()
      /Users/aneci/gits/argo-cd/server/application/application_test.go:2201 +0x1ec
  testing.tRunner()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1690 +0x184
  testing.(*T).Run.gowrap1()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1743 +0x40
```

adriananeci commented:
Raised https://github.com/argoproj/argo-cd/issues/21182 for tracking this race condition.
